File: server/generate_excel.js
import * as XLSX from 'xlsx';
import * as fs from 'fs';
import * as path from 'path';

function generateDatasets(outputDir) {
  if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir, { recursive: true });
  }

  const salesReps = [
    "Vikram Mehta", "Ananya Sharma", "Rahul Verma", "Priya Nair",
    "Rohan Deshmukh", "Neha Kapoor", "Amitabh Sen"
  ];

  const industries = [
    "Banking & Finance", "Healthcare & Life Sciences", "Manufacturing",
    "Retail & E-commerce", "Technology & SaaS", "Energy & Utilities", "Telecommunications"
  ];

  const solutions = [
    "Cloud Infrastructure", "Cybersecurity Suite", "Networking Architecture",
    "Enterprise ERP", "AI Data Platform", "Annual Maintenance Contract (AMC)"
  ];

  const leadSources = [
    "Direct Outreach", "Google Ads", "Partner Referral",
    "LinkedIn Campaign", "Industry Summit", "Inbound Website", "Cold Email"
  ];

  const customers = [
    "Apex Global Financial", "Zenith Health Systems", "Titan Heavy Motors", "Nova Retail Tech", "Starlight Software",
    "Aether Energy Corp", "Vanguard Telecom", "Omega Capital Solutions", "BioPharma Dynamics", "Precision Manufacturing",
    "Horizon Logistics", "Quantum Cloud Labs", "Metro Rail Systems", "Summit Securities", "Pulse MedTech",
    "Silverline Infra", "Hyperion Cyber Sec", "Nexus Data Labs", "Crest Insurance", "Orbital Satellite Tech"
  ];

  const lostReasons = [
    "Pricing / High Cost", "Competitor Selection (Better Features)", "Delayed Quotation / Slow Follow-up",
    "Budget Cut / Project Postponed", "Lack of Specific Feature", "Internal Restructuring", "Unresponsive Stakeholders"
  ];

  const stages = ["Qualification & Discovery", "Solution Architecture", "Commercial Proposal", "Contract Negotiation", "Final Approval"];

  // 1. WON DEALS
  const wonRows = [];
  for (let i = 1; i <= 140; i++) {
    const rep = salesReps[Math.floor(Math.random() * salesReps.length)];
    const ind = industries[Math.floor(Math.random() * industries.length)];
    const sol = solutions[Math.floor(Math.random() * solutions.length)];
    const src = leadSources[Math.floor(Math.random() * leadSources.length)];
    const cust = customers[Math.floor(Math.random() * customers.length)];

    const baseVal = [850000, 1500000, 2800000, 4500000, 7500000, 12000000, 25000000][Math.floor(Math.random() * 7)];
    const grossVal = baseVal + Math.floor(Math.random() * 150000);
    const netVal = Math.round((grossVal / 1.18) * 100) / 100;
    const gstVal = Math.round((grossVal - netVal) * 100) / 100;

    const month = Math.floor(Math.random() * 12) + 1;
    const day = Math.floor(Math.random() * 28) + 1;
    const dateStr = `2025-${String(month).padStart(2, '0')}-${String(day).padStart(2, '0')}`;

    wonRows.push({
      "Deal ID": `DEAL-WON-${1000 + i}`,
      "Customer Name": `  ${cust}  `, // extra space for cleaning test
      "Gross Revenue (INR)": grossVal,
      "GST (18%)": gstVal,
      "Net Revenue (INR)": netVal,
      "Sales Representative": rep,
      "Industry Vertical": ind,
      "Solution / Product": sol,
      "Lead Source Channel": src,
      "Close Date": dateStr,
      "Sales Cycle (Days)": Math.floor(Math.random() * 90) + 14,
      "Contract Term (Months)": [12, 24, 36, 48][Math.floor(Math.random() * 4)],
      "Margin %": Math.round((20 + Math.random() * 25) * 10) / 10
    });
  }

  // 2. LOST DEALS
  const lostRows = [];
  for (let i = 1; i <= 75; i++) {
    const rep = salesReps[Math.floor(Math.random() * salesReps.length)];
    const ind = industries[Math.floor(Math.random() * industries.length)];
    const sol = solutions[Math.floor(Math.random() * solutions.length)];
    const src = leadSources[Math.floor(Math.random() * leadSources.length)];
    const cust = customers[Math.floor(Math.random() * customers.length)];
    const reason = lostReasons[Math.floor(Math.random() * lostReasons.length)];

    const grossVal = [600000, 1200000, 2500000, 5000000, 9000000, 18000000][Math.floor(Math.random() * 6)];
    const month = Math.floor(Math.random() * 12) + 1;
    const day = Math.floor(Math.random() * 28) + 1;
    const dateStr = `2025-${String(month).padStart(2, '0')}-${String(day).padStart(2, '0')}`;

    lostRows.push({
      "Deal Reference ID": `DEAL-LOST-${2000 + i}`,
      "Client Organization": cust,
      "Quoted Gross Value": grossVal,
      "Estimated Net Loss": Math.round((grossVal / 1.18) * 100) / 100,
      "Sales Owner": rep,
      "Industry Sector": ind,
      "Proposed Solution": sol,
      "Acquisition Source": src,
      "Primary Lost Reason": reason,
      "Winning Competitor": ["Acme Enterprise", "TechCorp Global", "InnoSystems", "None / Internal"][Math.floor(Math.random() * 4)],
      "Lost Date": dateStr,
      "Sales Velocity Days": Math.floor(Math.random() * 80) + 10
    });
  }

  // 3. IN PROGRESS DEALS
  const progressRows = [];
  for (let i = 1; i <= 65; i++) {
    const rep = salesReps[Math.floor(Math.random() * salesReps.length)];
    const ind = industries[Math.floor(Math.random() * industries.length)];
    const sol = solutions[Math.floor(Math.random() * solutions.length)];
    const src = leadSources[Math.floor(Math.random() * leadSources.length)];
    const cust = customers[Math.floor(Math.random() * customers.length)];
    const stg = stages[Math.floor(Math.random() * stages.length)];
    const probMap = { "Qualification & Discovery": 20, "Solution Architecture": 40, "Commercial Proposal": 60, "Contract Negotiation": 80, "Final Approval": 90 };
    const prob = probMap[stg];

    const grossVal = [1000000, 2200000, 4800000, 8500000, 15000000, 30000000][Math.floor(Math.random() * 6)];
    const netVal = Math.round((grossVal / 1.18) * 100) / 100;
    const dateStr = `2025-09-15`;

    progressRows.push({
      "Opportunity ID": `DEAL-PIPE-${3000 + i}`,
      "Target Customer": cust,
      "Pipeline Gross Amount": grossVal,
      "Pipeline Net Amount": netVal,
      "Deal Owner": rep,
      "Industry": ind,
      "Solution Package": sol,
      "Lead Channel": src,
      "Current Pipeline Stage": stg,
      "Win Probability (%)": prob,
      "Weighted Forecast Net (INR)": Math.round(netVal * (prob / 100)),
      "Expected Close Date": dateStr,
      "Age in Pipeline (Days)": Math.floor(Math.random() * 100) + 5
    });
  }

  const wbWon = XLSX.utils.book_new();
  XLSX.utils.book_append_sheet(wbWon, XLSX.utils.json_to_sheet(wonRows), "Won Deals");
  XLSX.writeFile(wbWon, path.join(outputDir, "Won Deals.xlsx"));

  const wbLost = XLSX.utils.book_new();
  XLSX.utils.book_append_sheet(wbLost, XLSX.utils.json_to_sheet(lostRows), "Lost Deals");
  XLSX.writeFile(wbLost, path.join(outputDir, "Lost Deals.xlsx"));

  const wbProgress = XLSX.utils.book_new();
  XLSX.utils.book_append_sheet(wbProgress, XLSX.utils.json_to_sheet(progressRows), "In Progress Deals");
  XLSX.writeFile(wbProgress, path.join(outputDir, "In Progress Deals.xlsx"));

  console.log(`Generated 3 Excel datasets in ${outputDir}`);
}

generateDatasets("public/sample_data");
generateDatasets(".");

File: server/generate_excel.py
"""
Sample Excel Dataset Generator for Enterprise AI Sales Intelligence Platform
Generates 3 enterprise Excel files:
1. Won Deals.xlsx
2. Lost Deals.xlsx
3. In Progress Deals.xlsx
"""

import os
import pandas as pd
import numpy as np
from datetime import datetime, timedelta

def generate_datasets(output_dir="."):
    os.makedirs(output_dir, exist_ok=True)
    
    np.random.seed(42)
    
    sales_reps = [
        "Vikram Mehta", "Ananya Sharma", "Rahul Verma", "Priya Nair", 
        "Rohan Deshmukh", "Neha Kapoor", "Amitabh Sen"
    ]
    
    industries = [
        "Banking & Finance", "Healthcare & Life Sciences", "Manufacturing", 
        "Retail & E-commerce", "Technology & SaaS", "Energy & Utilities", "Telecommunications"
    ]
    
    solutions = [
        "Cloud Infrastructure", "Cybersecurity Suite", "Networking Architecture", 
        "Enterprise ERP", "AI Data Platform", "Annual Maintenance Contract (AMC)"
    ]
    
    lead_sources = [
        "Direct Outreach", "Google Ads", "Partner Referral", 
        "LinkedIn Campaign", "Industry Summit", "Inbound Website", "Cold Email"
    ]
    
    customers = [
        "Apex Global Financial", "Zenith Health Systems", "Titan Heavy Motors", "Nova Retail Tech", "Starlight Software",
        "Aether Energy Corp", "Vanguard Telecom", "Omega Capital Solutions", "BioPharma Dynamics", "Precision Manufacturing",
        "Horizon Logistics", "Quantum Cloud Labs", "Metro Rail Systems", "Summit Securities", "Pulse MedTech",
        "Silverline Infra", "Hyperion Cyber Sec", "Nexus Data Labs", "Crest Insurance", "Orbital Satellite Tech"
    ]
    
    lost_reasons = [
        "Pricing / High Cost", "Competitor Selection (Better Features)", "Delayed Quotation / Slow Follow-up",
        "Budget Cut / Project Postponed", "Lack of Specific Feature", "Internal Restructuring", "Unresponsive Stakeholders"
    ]
    
    stages = ["Qualification & Discovery", "Solution Architecture", "Commercial Proposal", "Contract Negotiation", "Final Approval"]
    
    start_date = datetime(2025, 1, 1)
    
    # 1. WON DEALS DATASET
    won_records = []
    for i in range(1, 140):
        rep = np.random.choice(sales_reps)
        ind = np.random.choice(industries)
        sol = np.random.choice(solutions)
        src = np.random.choice(lead_sources)
        cust = np.random.choice(customers)
        
        base_val = np.random.choice([850000, 1500000, 2800000, 4500000, 7500000, 12000000, 25000000])
        gross_val = float(base_val + np.random.randint(-50000, 100000))
        gst_amount = round(gross_val - (gross_val / 1.18), 2)
        net_val = round(gross_val / 1.18, 2)
        
        close_dt = start_date + timedelta(days=int(np.random.randint(0, 540)))
        sales_cycle = int(np.random.randint(14, 120))
        
        won_records.append({
            "Deal ID": f"DEAL-WON-{1000 + i}",
            "Customer Name": f"  {cust}  ", # extra space for cleaning test
            "Gross Revenue (INR)": gross_val,
            "GST (18%)": gst_amount,
            "Net Revenue (INR)": net_val,
            "Sales Representative": rep,
            "Industry Vertical": ind,
            "Solution / Product": sol,
            "Lead Source Channel": src,
            "Close Date": close_dt.strftime("%Y-%m-%d"),
            "Sales Cycle (Days)": sales_cycle,
            "Contract Term (Months)": np.random.choice([12, 24, 36, 48]),
            "Margin %": round(float(np.random.uniform(22.0, 48.0)), 1)
        })
        
    df_won = pd.DataFrame(won_records)
    
    # 2. LOST DEALS DATASET
    lost_records = []
    for i in range(1, 75):
        rep = np.random.choice(sales_reps)
        ind = np.random.choice(industries)
        sol = np.random.choice(solutions)
        src = np.random.choice(lead_sources)
        cust = np.random.choice(customers)
        reason = np.random.choice(lost_reasons, p=[0.38, 0.22, 0.15, 0.10, 0.08, 0.04, 0.03])
        
        gross_val = float(np.random.choice([600000, 1200000, 2500000, 5000000, 9000000, 18000000]))
        close_dt = start_date + timedelta(days=int(np.random.randint(0, 540)))
        competitor = np.random.choice(["Acme Enterprise", "TechCorp Global", "InnoSystems", "None / Internal", "Legacy Vendor"])
        
        lost_records.append({
            "Deal Reference ID": f"DEAL-LOST-{2000 + i}",
            "Client Organization": cust,
            "Quoted Gross Value": gross_val,
            "Estimated Net Loss": round(gross_val / 1.18, 2),
            "Sales Owner": rep,
            "Industry Sector": ind,
            "Proposed Solution": sol,
            "Acquisition Source": src,
            "Primary Lost Reason": reason,
            "Winning Competitor": competitor,
            "Lost Date": close_dt.strftime("%Y-%m-%d"),
            "Sales Velocity Days": np.random.randint(10, 90)
        })
        
    df_lost = pd.DataFrame(lost_records)
    
    # 3. IN PROGRESS DEALS DATASET
    progress_records = []
    for i in range(1, 65):
        rep = np.random.choice(sales_reps)
        ind = np.random.choice(industries)
        sol = np.random.choice(solutions)
        src = np.random.choice(lead_sources)
        cust = np.random.choice(customers)
        stg = np.random.choice(stages)
        prob = {"Qualification & Discovery": 20, "Solution Architecture": 40, "Commercial Proposal": 60, "Contract Negotiation": 80, "Final Approval": 90}[stg]
        
        gross_val = float(np.random.choice([1000000, 2200000, 4800000, 8500000, 15000000, 30000000]))
        est_close_dt = datetime.now() + timedelta(days=int(np.random.randint(5, 180)))
        
        progress_records.append({
            "Opportunity ID": f"DEAL-PIPE-{3000 + i}",
            "Target Customer": cust,
            "Pipeline Gross Amount": gross_val,
            "Pipeline Net Amount": round(gross_val / 1.18, 2),
            "Deal Owner": rep,
            "Industry": ind,
            "Solution Package": sol,
            "Lead Channel": src,
            "Current Pipeline Stage": stg,
            "Win Probability (%)": prob,
            "Weighted Forecast Net (INR)": round((gross_val / 1.18) * (prob / 100.0), 2),
            "Expected Close Date": est_close_dt.strftime("%Y-%m-%d"),
            "Age in Pipeline (Days)": np.random.randint(5, 120)
        })
        
    df_progress = pd.DataFrame(progress_records)
    
    won_path = os.path.join(output_dir, "Won Deals.xlsx")
    lost_path = os.path.join(output_dir, "Lost Deals.xlsx")
    progress_path = os.path.join(output_dir, "In Progress Deals.xlsx")
    
    df_won.to_excel(won_path, index=False, sheet_name="Won Deals")
    df_lost.to_excel(lost_path, index=False, sheet_name="Lost Deals")
    df_progress.to_excel(progress_path, index=False, sheet_name="In Progress Deals")
    
    print(f"Generated Excel datasets:\n - {won_path}\n - {lost_path}\n - {progress_path}")

if __name__ == "__main__":
    generate_datasets("public/sample_data")
    generate_datasets(".")
